test/cpp/end2end/xds/xds_utils.h

```diff
@@ -43,7 +43,13 @@ class XdsBootstrapBuilder {
   // not already been set.
   XdsBootstrapBuilder& SetDefaultServer(const std::string& server,
                                         bool ignore_if_set = false) {
-    if (!ignore_if_set || top_server_.empty()) top_server_ = server;
+    if (!ignore_if_set || absl::c_find(servers_, server) == servers_.end()) {
```

Comment:
The `ignore_if_set` thing was a bit of a hack to allow tests to override the default server.  The idea is that the test framework [automatically sets the default server, but only if it was not already set by the application](https://github.com/grpc/grpc/blob/4ce4f4ea3aa2fd5863b47cfa2eee4de7cdd6306d/test/cpp/end2end/xds/xds_end2end_test_lib.cc#L504).  Given that pattern, I think we would want `ignore_if_set` to ignore the value if the current list is non-empty.  It should not matter whether the specified server is in the list or not, just that the list has already been populated.

Actually, I think it might be cleaner to just eliminate this `ignore_if_set` hack completely.  Instead, I suggest the following:

- In `XdsEnd2endTest`, add this method:
  ```
  XdsBootstrapBuilder MakeBootstrapBuilder() {
    return XdsBootstrapBuilder().SetServer(
        absl::StrCat("localhost:", balancer_->port()));
  }
  ```
- In all tests where we directly instantiate an `XdsBootstrapBuilder` but do not set the server, we should instead call `MakeBootstrapBuilder()`, so that the default server is set automatically.
- In cases where we instantiate an `XdsBootstrapBuilder` and *do* override the default server (it looks like this we do this in [only one place](https://github.com/grpc/grpc/blob/4ce4f4ea3aa2fd5863b47cfa2eee4de7cdd6306d/test/cpp/end2end/xds/xds_core_end2end_test.cc#L1283)), the code should continue to do what it's already doing.
- We can remove [the line in `InitClient()` that sets the default server](https://github.com/grpc/grpc/blob/4ce4f4ea3aa2fd5863b47cfa2eee4de7cdd6306d/test/cpp/end2end/xds/xds_end2end_test_lib.cc#L504).

Reply:
I tried to contain the blast radius, hence the use of `absl::optional`. Please take another look.